keylime/config.py

```diff
@@ -73,45 +74,257 @@ def environ_bool(env_name, default):
+            raise Exception(f"Invalid component {component}")
+
         # TODO - use logger and be sure that all variables have a
         # propper default, and the sections are initialized
-        if not any(os.path.exists(c) for c in CONFIG_FILES):
-            print(
-                f"Config file not found in {CONFIG_FILES}. Please set "
-                f"environment variable KEYLIME_CONFIG or see {__file__} "
-                "for more details."
-            )
+        if not any(os.path.exists(c) for c in CONFIG_FILES[component]):
+            print(f"Config file not found in {CONFIG_FILES[component]}. " f"Please see {__file__} for more details.")
             if _CONFIG_FALLBACK:
                 print("Falling back on package provided configuration")
-                file = resources.files(__package__).joinpath("keylime.conf")
-                config.read_string(file.read_text("utf-8"))
+                file = resources.files(__package__).joinpath(f"config/{component}.conf")
+                _config[component].read_string(file.read_text("utf-8"))
         else:
-            # Validate that at least one config file is present
-            config_files = config.read(CONFIG_FILES)
-            # TODO - use the logger
-            print(f"Reading configuration from {config_files}")
-        get_config.config = config
-    return get_config.config
+            for c in CONFIG_FILES[component]:
+                # Search for configuration file in order of priority given by
+                # CONFIG_FILES. The first base configuration file found is used,
+                # the others are ignored
+
+                # Validate that at least one config file is present
+                config_file = _config[component].read(c)
+                # TODO - use the logger
+                if config_file:
+                    print(f"Reading configuration from {config_file}")
+
+                    if not CONFIG_SNIPPETS_DIRS or not isinstance(CONFIG_SNIPPETS_DIRS, dict):
+                        raise Exception("Invalid CONFIG_FILES")
+
+                    if not component in CONFIG_SNIPPETS_DIRS:
+                        raise Exception(f"Invalid component {component}")
+
+                    for d in (x for x in CONFIG_SNIPPETS_DIRS[component] if os.path.exists(x)):
+                        snippets = sorted(filter(os.path.isfile, (os.path.join(d, f) for f in os.listdir(d) if f)))
+                        applied_snippets = _config[component].read(snippets)
+                        if applied_snippets:
+                            # TODO - use the logger
+                            print(f"Applied configuration snippets from {d}")
+                    break
+
+    return _config[component]
+
+
+def getlist(component, option, section=None):
+    if not section:
+        section = component
+
+    read = get_config(component).get(section, option)
+
+    if read:
+        try:
+            l = ast.literal_eval(read)
```

Comment:
I was wondering if this is safe or not. Seems that it is, as is a very focused evaluator. Maybe we should add a note.

Reply:
The documentation states it is safe. It never executes the code, only parses a restrict subset of literal values.

See: https://docs.python.org/3/library/ast.html#ast.literal_eval